File: 2016/03/03.py
from itertools import product

def parser(data: str): 
    return list(map(int, data.split()))

def solver(lengths):
    yield count_possible_triangles(lengths[i: i + 3] for i in range(0, len(lengths), 3))
    vertical = ((lengths[i + j],     \
                 lengths[i + j + 3], \
                 lengths[i + j + 6]) for i, j in product(range(0, len(lengths), 9), range(3)))
    yield count_possible_triangles(vertical)

def count_possible_triangles(triangles):
    return sum(sum(triangle) > 2 * max(triangle) for triangle in triangles)

File: 2016/03/03.rs
aoc::main!();

fn preprocessing(input: &str) -> Vec<u16> {
    input
    .split([' ', '\n'])
    .filter_map(|n| n.parse::<u16>().ok())
    .collect_vec()
}

fn part_1(lengths : Vec<u16>) -> usize {
    solver(lengths, 1)
}

fn part_2(lengths : Vec<u16>) -> usize {
    solver(lengths, 3)
}


fn solver(lengths : Vec<u16>, shift: usize) -> usize {
    (0..lengths.len() / 3)
    .map(|t| ((t % shift) + (t / shift) * (3 * shift)))
    .map(|f| [lengths[f] , lengths[f + shift], lengths[f + 2 * shift]])
    .filter(|v| v.iter().sum::<u16>() > 2 * v.iter().max().unwrap())
    .count()
}